sdk/core/azure-core/src/http/curl/curl_connection_private.hpp

```diff
@@ -169,9 +169,10 @@ namespace Azure { namespace Core { namespace Http {
 #endif
         if (result != CURLE_OK)
         {
-          throw Http::TransportException(
+          throw Http::CurlTransportException(
```

Comment:
IIRC, @antkmsft was one who proposed to get rid of all the sub-types for TransportExceptions and just use the one that we have.  Do you remember Anton?

Reply:
As I mentioned above: We have a feature request from the cognitive services folk to expose the low level error from the transport in the exception returned to the caller.

My solution for this was to add a transport specific exception type and push the problem to the cognitive services team.

If you have an alternate solution that achieves the same goal I'm completely open to the solution. Remember that transport errors are domain specific, so the caller needs to be able to know which transport created the error code (and not all transports will report errors as error codes).

Reply:
makes sense.
Here's what I think:
If service team needs it, that means that the sdk client will need to know
about what transport adapter is in the options. Some thoughts:
- On sdk client creating, we will need to check whats the transport and save it. This way, when calling sdk operations, we will know what to catch?
- If we don't save that's the adapter type, the operation will need to try catching all types of exception.
- For custom adapter, there won't be a code in the exception.

I would try to understand a little better why do they need the low level error code.  Usually a service team would only care about the HTTP error code, or some other standard error code.  What can the sdk client implementation do with a low level error?  If curl returns error code = XX, I don't see how that can be useful for the sdk client.

But, if they really need it, I would suggest making it part of the transport adapter contract, so all transport adapters would implement it.
This means making a default implementation for the Send() function like:
```cpp
Send() {
   try {
       SendImpl(..,..,..)
  } catch (transportEx e) {
     m_lowLevelErrorCode = GetErrorCodeImpl()
}
```

All transport adapter would need to implement SendImpl() and GetErrorCodeImp()

Reply:
The problem is that the type of `m_lowLevelErrorCode` is dependent on the transport. For CURL, it's an `int`, for WinHTTP, it's a `DWORD`. For some other transport, we don't know what it is - it could even be a structured type or a `std::exception`.

This solution punts the problem to the caller. And it's possible to write code which handles this in a meaningful way without knowing which transport was selected:
```c++
try { ... }
catch (CurlTransportException &cte)
{
 // Process CURL transport exception.
}
catch (WinHttpTransportException &wte)
{
  // Process WinHTTP transport exception.
}
catch (TransportException &te)
{
// Fallback
}
```

Not a great story, but it meets the cognitive requirements as best as is possible given the constraints.

Reply:
@vhvb1989  I don't remember, but in general I see it and I think that maybe it would be better to try to avoid per-transport exceptions and options.

Reply:
@LarryOsterman, you probably won't be able to write that try-catch that easily, unless you compiled SDK with both libcurl and WinHTTP support, because transport-specific types (plus DWORDs etc ) are only available when the SDK was built with a specific flag.

Reply:
The scenario Victor called out was a customer who (a) needed to know the HTTP transport and (b) compiled the SDK with both libcurl and WinHTTP. 

I am eagerly anticipating your solution which allows me to meet the customer requirement that I provide the error code returned from the underlying transport which also respects the fact that the error information is transport specific (and our two existing transports have different underlying types of error codes).

Reply:
try { ... }
catch (DetailedTransportException<CURLcode> &cte)
{
 // Process CURL transport exception.
}
catch (DetailedTransportException<DWORD> &wte)
{
  // Process WinHTTP transport exception.
}
catch (TransportException &te)
{
// Fallback
}

It requires the definition of types named "CURLcode" and "DWORD", but I believe those can be defined without the entire transport header. 

Again, the scenario here is that the caller *knows* which transports may be used and how to interpret the error codes from the transport (that's an implicit requirement because there is no mapping between Win32 WinHTTP error codes and CURL CURLcode values).

Reply:
Curious...
If we are creating a common abstraction for Proxy across curl and winhttp, why not creating a common abstraction for an error code returned?

the Proxy common abstraction is translating the proxy string to the specifics for each transport.
We could also translate the error codes to be the same?

Reply:
The low level error information is inherently specific to the transport. There are dozens and dozens of WIn32 errors that might be returned, and dozens of CURL errors that might be returned. It's not feasible to maintain those mappings over time and retain the semantics.

You can think of this as being similar to [errno and _doserrno](https://github.com/MicrosoftDocs/cpp-docs/blob/main/docs/c-runtime-library/errno-doserrno-sys-errlist-and-sys-nerr.md). The MS C runtime library provides two different "error" abstractions. One is standards based `errno` and the other is OS specific `_doserrno`. 

For customers that *need* the DOS error code, it's available.

This is exactly the same for CURL and WinHTTP.

Reply:
yes, what I meant is, similar to the Proxy, do we want a common error code abstraction?

We can set Proxy from the common options, but if we want something more specific, we need to go and create the transport and set options directly from there.

Now, for the error code..  there are also some common errors while connecting/reading/uploading.  For others, you we would go to the transport...

No asking to do it...  just curious...  Wondering only :)

Reply:
It's an interesting idea, and I think it might make sense. Let me ask the Cognitive folks tomorrow at our sync-up to confirm that they really need the low level error code, or if a proxy to the low level error code makes sense (this is especially true given the significant number of cases where we *don't* have a low level error code to return).

Reply:
Talking to the cognitive services folks, it appears that this feature request didn't have the behaviors that they wanted, backing this change out and revisiting in the future.